patches/server/0344-Generator-Settings.patch

```diff
@@ -0,0 +1,176 @@
+     @Override
+     public void buildSurface(WorldGenRegion region, StructureFeatureManager structures, ChunkAccess chunk) {
+         if (!SharedConstants.debugVoidTerrain(chunk.getPos())) {
+-            WorldGenerationContext worldgenerationcontext = new WorldGenerationContext(this, region);
++            WorldGenerationContext worldgenerationcontext = new WorldGenerationContext(this, region, structures.getWorld()); // Paper
+             NoiseGeneratorSettings generatorsettingbase = (NoiseGeneratorSettings) this.settings.get();
+             NoiseChunk noisechunk = chunk.getOrCreateNoiseChunk(this.sampler, () -> {
+                 return new Beardifier(structures, chunk);
+@@ -253,7 +253,7 @@ public final class NoiseBasedChunkGenerator extends ChunkGenerator {
+             return new Beardifier(structureAccessor, chunk);
+         }, (NoiseGeneratorSettings) this.settings.get(), this.globalFluidPicker, Blender.of(chunkRegion));
+         Aquifer aquifer = noisechunk.aquifer();
+-        CarvingContext carvingcontext = new CarvingContext(this, chunkRegion.registryAccess(), chunk.getHeightAccessorForGeneration(), noisechunk);
++        CarvingContext carvingcontext = new CarvingContext(this, chunkRegion.registryAccess(), chunk.getHeightAccessorForGeneration(), noisechunk, structureAccessor.getWorld()); // Paper
+         CarvingMask carvingmask = ((ProtoChunk) chunk).getOrCreateCarvingMask(generationStep);
+ 
+         for (int j = -8; j <= 8; ++j) {
+diff --git a/src/main/java/net/minecraft/world/level/levelgen/WorldGenerationContext.java b/src/main/java/net/minecraft/world/level/levelgen/WorldGenerationContext.java
+index b99283c31193e2110f6e3f39c23dbfc2442bab2b..5286fb50c0c65f2d135551e22ed6fd5e2db0d0b3 100644
+--- a/src/main/java/net/minecraft/world/level/levelgen/WorldGenerationContext.java
++++ b/src/main/java/net/minecraft/world/level/levelgen/WorldGenerationContext.java
+@@ -6,10 +6,13 @@ import net.minecraft.world.level.chunk.ChunkGenerator;
+ public class WorldGenerationContext {
+     private final int minY;
+     private final int height;
++    private final net.minecraft.world.level.Level level; // Paper
+ 
+-    public WorldGenerationContext(ChunkGenerator generator, LevelHeightAccessor world) {
++    public WorldGenerationContext(ChunkGenerator generator, LevelHeightAccessor world) { this(generator, world, null); } // Paper
++    public WorldGenerationContext(ChunkGenerator generator, LevelHeightAccessor world, net.minecraft.world.level.Level level) { // Paper
+         this.minY = Math.max(world.getMinBuildHeight(), generator.getMinY());
+         this.height = Math.min(world.getHeight(), generator.getGenDepth());
++        this.level = level; // Paper
+     }
+ 
+     public int getMinGenY() {
+@@ -19,4 +22,6 @@ public class WorldGenerationContext {
+     public int getGenDepth() {
+         return this.height;
+     }
++
++    public net.minecraft.world.level.Level getWorld() { return this.level; } // Paper
```

Comment:
Throw an exception that's not an NPE if this is called when level is null I think. I suppose its possible if some plugin creates a WorldGenerationContext